### DISCUSSION TITLE
[Detector Support]: Do frigate models utilize apple's neural engine cores?

### DISCUSSION THREAD

**Discussion by stas-milev** [blakeblackshear/frigate#10665] [Detector Support]
### Describe the problem you are having

I wonder if anyone ran "asitop" on M1/M2/M3 macs to check if frigate object detection models make any use of apple silicone's neural engine cores or whether the models need to be additionally converted to .mlmodel format to make use of them. I wonder how they compare against the coral devices provided neural cores can perform at 15-17 TOPS.

### Version

N/A

### Frigate config file

```yaml
N/A
```

### docker-compose file or Docker CLI command

```yaml
N/A
```

### Relevant log output

```shell
N/A
```

### Operating system

Other

### Install method

Docker CLI

### Coral version

Other

**Comment by NickM-27** [comment]
yes they are used, I use a macbook pro for my contributions to frigate and inference times are higher than coral but just by a few ms

**Reply by stas-milev** [reply]
thanks @NickM-27. it does not sound like a huge boost over a usb coral tpu but may end up adding up if youy add more cameras. Did you try converting the tf lite model to .mlmodel? it can help with compatibility. Will frigate accept it a a detector?

**Reply by bluearcher-bc** [reply]
i feel like a $599 m4 mac mini might be the best value frigate powerhouse on the market right now... am I wrong?

**Reply by stas-milev** [reply]
if only I knew the answer. :-) Docker does not support devices passthrough. you can't use Mac's GPU cores and you can't use a coral. Well there's a way of sharing a coral now (usbip) but some people say it's unreliable on a Mac. @NickM-27 did you try coral on a Mac via usbip server?

**Reply by NickM-27** [reply]
I have not tried the coral yet, Mac works very well already though

**Reply by stas-milev** [reply]
@NickM-27 do you run frigate docker container directly on your mac or do run it in a virualbox vm or alike?

**Reply by NickM-27** [reply]
Directly on Mac for Frigate development

**Reply by chanchiem** [reply]
@NickM-27 are there any guides or suggestions on how we can easily run this standalone on a Mac? It'd be nice to see what the performance difference is with hardware acceleration

**Reply by stas-milev** [reply]
I think what @NickM-27 meant was running up a docker on Mac (i.e. not native compilation for Mac OS).

**Reply by NickM-27** [reply]
Correct, running without docker is not supported.

**Reply by mr-ransel** [reply]
Hey @NickM-27 , Maybe I'm misunderstanding, but it sounds like you're saying that running frigate in a docker container on MacOS is successfully leveraging the neural cores on apple silicon for object detection? I'd previously thought that wasn't possible since docker on mac didn't allow it? (nor does it allow surfacing GPUs to containers, preventing hardware acceleration as well)

Can you share a bit more about what configuration you're using to do this? I'm trying to figure out how far I can push getting frigate to run at speed on an M2 or M3, and maximize hardware acceleration with tricks like using codeproject.ai locally which can run on the baremetal and have full access to hardware acceleration. Or similarly running go2rtc on the baremetal to preprocess all the streams into the desired format to exploit local hardware acceleration there as well. 

Hoping to learn a bit more to see how far I can take hardware acceleration and neural engine usage on apple silicon.

**Reply by NickM-27** [reply]
@mr-ransel No, it doesn't use the neural cores, but macos has tensorflow optimizations built in even on the CPU that allows it to run these inferences efficiently.

It's worth mentioning I already have a POC for using a local ipc proxy to use the neural cores on the host in python and have frigate communicate with that, running yolov9

**Reply by mr-ransel** [reply]
That's exciting news! Any chance we've got a future with more than a POC going upstream? I've got an Mac Mini M2 that just showed up on my doorstep I was going to put scrypted NVR on if you're interested in a guinea pig for more in-depth performance and stability testing.

**Reply by NickM-27** [reply]
Not sure what you mean by more than a POC, POC is simply the first step. Anyway, the feature has been merged into 0.17

https://github.com/blakeblackshear/frigate/pull/19592

**Reply by mr-ransel** [reply]
Oh sweet! Merging the feature for release was exactly what I meant! Glad to see it, and I'll experiment!

**Reply by gavc1** [reply]
Could this be used instead of docker for hosting and direct access to the hw?
https://github.com/apple/container

**Reply by NickM-27** [reply]
No, Apple container doesn't support GPU access
